Fix parameter types and other fixes

dagwieers opened the pull request:
##### SUMMARY
This PR adds parameter types to files and system modules.

This PR also includes:
- Fixes to arg_spec to add type
- Add missing trailing dots in descriptions
- Remove unwanted trailing dots in short_description
- Improve parameter descriptions

##### ISSUE TYPE
- Docs Pull Request
- Feature Pull Request

##### COMPONENT NAME
files and system modules

dagwieers committed:
Fix parameter types and other fixes

acozine commented on lib/ansible/modules/files/copy.py:
```diff
@@ -174,17 +180,17 @@
     dest: /etc/mine.conf
 
-- name: if follow is true, /path/to/file will be overwritten by contents of foo.conf
+- name: If follow is true, /path/to/file will be overwritten by contents of foo.conf
```
I'd update the `name` line to match the parameter. So "If follow is set to 'yes', . . . ", since we have `follow: yes` below.

dagwieers replied:
Sure, and I have been doing that for some, but it was not my intention to review everything as that would take too much time :-)

dagwieers replied:
I mostly fix what I find on my path towards fixing the parameter types.

acozine reviewed (commented):
GitHub seems to be stuck. I'm going to see if it will let me add this comment . . . I do have a few more.

dagwieers committed:
Fix issues after review

dagwieers committed:
Fix Windows-references in system/files modules

This PR includes:
- Replacing version/v with just Ansible X.Y
- Removing Windows-alternatives from notes

acozine commented on lib/ansible/modules/system/authorized_key.py:
```diff
@@ -13,81 +13,86 @@
-      - Alternate path to the authorized_keys file
-    default: "(homedir)+/.ssh/authorized_keys"
+      - Alternate path to the authorized_keys file.
+      - When unset, this value defaults to I(~/.ssh/authorized_keys).
```
If I were scanning the docs for this parameter, looking for a default, I don't think I would see this. Everywhere else in the docs we list default values in a separate line with the label `default:`. Is there a reason this one should be different?

dagwieers replied:
Yes, two reasons. It's not a module default, but a low-level tool default, and it causes a sanity error because the arg_spec doesn't match the documentation.

We have various other cases where defaults cannot be put as a default in the docs and are listed in the description, so it's not that exceptional.

acozine replied:
That makes sense, thanks for the explanation.

acozine commented on lib/ansible/modules/files/file.py:
```diff
@@ -111,21 +116,19 @@
     group: foo
-    # When specifying mode using octal numbers, add a leading 0
-    mode: 0644
+    mode: '0644'
```
Should this be `'644'` instead of `'0644'`?

dagwieers replied:
Nope, both are fine, but '0644' is preferable as it indicates an octal number and not the number 644.

acozine commented on lib/ansible/modules/files/file.py:
```diff
@@ -169,7 +172,7 @@
     path: /etc/some_directory
     state: directory
-    mode: 0755
+    mode: '0755'
```
Same as above - should we drop the leading zero when we add the quotes? The shared docs for the `mode` parameter say: 
```
For those used to /usr/bin/chmod remember that modes are actually octal numbers. 
You must either add a leading zero so that Ansible's YAML parser knows it is an octal number (like 0644 or 01777) 
or quote it (like '644' or '1777') so Ansible receives a string and can do its own conversion from string into number.
```

dagwieers replied:
Well, the 0 is not longer essential for it to work (while it was if there were no quotes) but the leading zero is essential to understand it is an octal number. So personally, I would avoid any confusion by leaving it in place for the examples, force strings and don't mention anything else.

If we force strings (#50035) we avoid any issues and if the zero is left out (by accident) it would still work.

acozine replied:
Okay, let's leave the leading zeros on the examples. And perhaps update the shared docs for `mode` again, depending on what happens with #50035.

acozine commented on lib/ansible/modules/files/ini_file.py:
```diff
@@ -8,99 +8,103 @@
     option: fav
     value: lemonade
-    mode: 0600
+    mode: '0600'
```
Should it be `'600'`?

dagwieers replied:
Nope.

acozine commented on lib/ansible/modules/files/lineinfile.py:
```diff
@@ -146,16 +155,16 @@
     owner: root
     group: root
-    mode: 0644
+    mode: '0644'
```
Here too - `'644'`?

dagwieers replied:
Nope :-)

acozine commented on lib/ansible/modules/files/replace.py:
```diff
@@ -31,88 +31,95 @@
     backup: yes
 
-# Replace after the expression till the end of the file (requires >=2.4)
+# Replace after the expression till the end of the file (requires Ansible >= 2.4)
```
```suggestion
- name: Replace after the expression till the end of the file (requires Ansible >= 2.4)
```

acozine commented on lib/ansible/modules/files/replace.py:
```diff
@@ -31,88 +31,95 @@
 
-# Replace after the expression till the end of the file (requires >=2.4)
+# Replace after the expression till the end of the file (requires Ansible >= 2.4)
 - replace:
```
```suggestion
  replace:
```

acozine commented on lib/ansible/modules/files/replace.py:
```diff
@@ -31,88 +31,95 @@
     backup: yes
 
-# Replace before the expression till the begin of the file (requires >=2.4)
+# Replace before the expression till the begin of the file (requires Ansible >= 2.4)
```
```suggestion
- name: Replace before the expression till the begin of the file (requires Ansible >= 2.4)
```

acozine commented on lib/ansible/modules/files/replace.py:
```diff
@@ -31,88 +31,95 @@
 
-# Replace before the expression till the begin of the file (requires >=2.4)
+# Replace before the expression till the begin of the file (requires Ansible >= 2.4)
 - replace:
```
```suggestion
  replace:
```

acozine commented on lib/ansible/modules/files/replace.py:
```diff
@@ -31,88 +31,95 @@
     backup: yes
 
-# Replace between the expressions (requires >=2.4)
+# Replace between the expressions (requires Ansible >= 2.4)
```
```suggestion
- name: Replace between the expressions (requires Ansible >= 2.4)
```

acozine commented on lib/ansible/modules/files/replace.py:
```diff
@@ -31,88 +31,95 @@
 
-# Replace between the expressions (requires >=2.4)
+# Replace between the expressions (requires Ansible >= 2.4)
 - replace:
```
```suggestion
  replace:
```

acozine commented on lib/ansible/modules/files/replace.py:
```diff
@@ -126,23 +133,23 @@
     owner: jdoe
     group: jdoe
-    mode: 0644
+    mode: '0644'
```
`'644'`?

acozine commented on lib/ansible/modules/net_tools/basics/get_url.py:
```diff
@@ -171,7 +179,7 @@
     url: http://example.com/path/file.conf
     dest: /etc/foo.conf
-    mode: 0440
+    mode: '0440'
```
`'440'`?

dagwieers replied:
Nope :-)

acozine commented on lib/ansible/modules/system/aix_lvol.py:
```diff
@@ -205,7 +216,7 @@ def main():
             size=dict(type='str'),
             opts=dict(type='str', default=''),
-            copies=dict(type='str', default='1'),
+            copies=dict(type='int', default=1),
```
Could this change break playbooks?

dagwieers replied:
No, because the value is used as a string in the code, and the example was using an integer. In fact it is an integer (and not just a string) so I'd rather have the module fail with a proper error that an integer is expected, than the command fail with a dubious error because something non-integer like was added.

However, I do understand if you would prefer to keep it out of this PR.

acozine replied:
Ah, excellent. I do think it would be better in a separate PR, since it's a different kind of fix. Pulling it out will make this change easier to find.

dagwieers replied:
Removed.

acozine commented on lib/ansible/modules/system/parted.py:
```diff
@@ -13,93 +13,97 @@
-     - If to create or delete a partition. If set to C(info) the module will
-       only return the device information.
-    choices: ['present', 'absent', 'info']
+    - If to create or delete a partition.
```
```suggestion
    - Whether to create or delete a partition.
```

acozine commented on lib/ansible/modules/system/service.py:
```diff
@@ -24,51 +24,58 @@
-          behaving init scripts that exit immediately after signaling a process
-          to stop.
+          between the stop and start command.
+        - This helps to workaround badly behaving init scripts that exit immediately
```
```suggestion
        - This helps to work around badly-behaving init scripts that exit immediately
```

acozine commented on lib/ansible/modules/system/service.py:
```diff
@@ -24,51 +24,58 @@
         description:
-        - "For OpenRC init scripts (ex: Gentoo) only.  The runlevel that this service belongs to."
+        - For OpenRC init scripts (e.g. Gentoo) only.
+        - The runlevel that this service belongs to."
```
```suggestion
        - The runlevel that this service belongs to.
```

acozine reviewed (commented):
My comments from yesterday got clobbered somehow, so I've re-reviewed - some of these comments were discussed on IRC, others were not.

acozine committed:
Update lib/ansible/modules/system/parted.py

Co-Authored-By: dagwieers <dag@wieers.com>

acozine committed:
Update lib/ansible/modules/system/service.py

Co-Authored-By: dagwieers <dag@wieers.com>

acozine committed:
Update lib/ansible/modules/system/service.py

Co-Authored-By: dagwieers <dag@wieers.com>

dagwieers committed:
Revert type change, move to separate PR

acozine committed:
Update lib/ansible/modules/files/replace.py

Co-Authored-By: dagwieers <dag@wieers.com>

acozine committed:
Update lib/ansible/modules/files/replace.py

Co-Authored-By: dagwieers <dag@wieers.com>

acozine committed:
Update lib/ansible/modules/files/replace.py

Co-Authored-By: dagwieers <dag@wieers.com>

acozine committed:
Update lib/ansible/modules/files/replace.py

Co-Authored-By: dagwieers <dag@wieers.com>

acozine committed:
Update lib/ansible/modules/files/replace.py

Co-Authored-By: dagwieers <dag@wieers.com>

acozine committed:
Update lib/ansible/modules/files/replace.py

Co-Authored-By: dagwieers <dag@wieers.com>

dagwieers commented:
A rebase was required after merging #50602

Please re-review and merge !

dagwieers commented:
I add to do another rebase for a new conflict. Please re-review.

acozine commented on lib/ansible/modules/system/pamd.py:
```diff
@@ -13,93 +13,93 @@
+    - If state is C(args_absent) or C(args_present), new_type, new_control, and new_module_path will be ignored.
+    - State C(absent) will remove the rule.  The 'absent' state was added in Ansible 2.4.
+    type: str
+    choices: [ absent, before, after, args_absent, args_present, updated ]
```
I'm surprised you didn't alphabetize the choices . . . I'm fine with merging this as-is, or changing it, just thought it might bother you in future.

dagwieers replied:
```suggestion
    choices: [ absent, after, args_absent, args_present, before, updated ]
```
Haha, I would have done that, but sometimes I can restrain myself from doing more than intended. This was such a time I guess :-P

dagwieers commented:
Another rebase was required. Can we please get this merged?

gundalow commented:
rebuild_merge
